hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/IterableStorageManager.java

```diff
@@ -51,29 +61,200 @@ public IterableStorageManager() {
+     * @param enhancement the enhancement for the current transaction
+     * @param contractNumber the contract number
+     * @return the first storage key for the contract or null if none exists.
+     */
+    @Nullable
+    private Bytes contractFirstKeyOf(@NonNull final Enhancement enhancement, long contractNumber) {
+        final var account = enhancement.nativeOperations().getAccount(contractNumber);
+        return account != null && account.firstContractStorageKey() != null ? account.firstContractStorageKey() : null;
+    }
+
+    /**
+     * Removes the given key from the slot storage and from the linked list of storage for the given contract, and removes the
+     *
+     * @param store Contract storage store
+     * @param firstContractKey The first key in the linked list of storage for the given contract
+     * @param contractNumber The contract number under consideration
+     * @param key The slot key to remove
+     * @return the new first key in the linked list of storage for the given contract
+     */
+    @Nullable
+    private Bytes removeAccessedValue(
+            @NonNull final ContractStateStore store,
+            @Nullable final Bytes firstContractKey,
+            long contractNumber,
+            @NonNull final Bytes key) {
+        try {
+            Objects.requireNonNull(store);
+            Objects.requireNonNull(key);
+            final var slotKey = newSlotKeyFor(contractNumber, key);
+            final var slotValue = newSlotValueFor(store, false, slotKey, "Missing key ", key);
+            final var nextKey = slotValue.nextKey();
+            final var prevKey = slotValue.previousKey();
+
+            if (nextKey != null) {
+                // Look up the next slot value
+                final var nextSlotKey = newSlotKeyFor(contractNumber, nextKey);
+                final var nextValue = newSlotValueFor(store, true, nextSlotKey, "Missing next key ", nextKey);
+
+                // Create new next value and put into the store
+                final var newNextValue =
+                        nextValue.copyBuilder().previousKey(prevKey).build();
+                store.putSlot(nextSlotKey, newNextValue);
+            }
+            if (prevKey != null) {
+                // Look up the previous slot value
+                final var prevSlotKey = newSlotKeyFor(contractNumber, prevKey);
+                final var prevValue = newSlotValueFor(store, true, prevSlotKey, "Missing previous key ", prevKey);
+
+                // Create new previous value and put into the store
+                final var newPrevValue =
+                        prevValue.copyBuilder().nextKey(nextKey).build();
+                store.putSlot(prevSlotKey, newPrevValue);
+            }
+            store.removeSlot(slotKey);
+            return key.equals(firstContractKey) ? slotValue.nextKey() : firstContractKey;
+        } catch (Exception irreparable) {
+            log.error(
```

Comment:
Just occurred to me: Here (+ update + insert) should the exception be rethrown?  (Or wrapped in another exception and thrown?)

If so, two issues occur at first glance:
a) would caller `persistChanges` let it fly through (stopping then at first error) or would it collect possibly multiple such exceptions and wrap them together and throw a single exception for a batch?  (Way overkill I imagine)
b) way upstream (I think) is `TransactionProcessor.safeCommit` which _does_ handle a `ResourceExhaustedException` specifically but no other.  So I don't know _what_ happens if a different exception is thrown when trying to commit.

Reply:
Mono service does not seem to re-throw the exception.  Doesn't mean that is correct.  The way things are now, an exception in one change does not affect the processing of the rest of the changes which might be the right behavior but certainly open for discussion.